### PR TITLE
Handle deletion of uncommitted news fragments

### PR DESCRIPTION
**Pull request by abompard**
Before this commit, all the news fragments needed to be committed into git, or the fragments removal after building the news file would crash.

In my workflow, I add missing fragments before building the news file because I'm extracting author names from the git log, and towncrier crashes at the end of the build process.

### CODE REVIEW

File: src/towncrier/newsfragments/357.feature.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Handle deletion of uncommitted news fragments
```
Comment (adiroiban) [inline src/towncrier/newsfragments/357.feature.rst:1]: Maybe something like this.

```suggestion
It can now handle news fragments which are not part of the git repository. For example uncommited or unstagged files.
```

File: src/towncrier/test/test_build.py [hunk 2]
```diff
@@ -836,3 +836,81 @@ def test_start_string(self):
         """)
 
         self.assertEqual(expected_output, output)
+
+    def test_uncommitted_files(self):
+        runner = CliRunner()
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py:841]: It would be nice to have a dostring describing what feature this test tries to cover.

```suggestion

    def test_uncommitted_files(self):
        """
        At build time, it will delete any fragment file regardless of its stage,
        included files that are not part of the git reporsitory,
        or are just staged or modified.
        """
        runner = CliRunner()
```

File: src/towncrier/test/test_build.py [hunk 3]
```diff
@@ -836,3 +836,81 @@ def test_start_string(self):
...
+                    --------
+
+                    - Adds levitation (#123)
+                    - Extends levitation for an hour (#124)
+                    - Baz levitation (#125)
+                    - Fix (literal) crash (#126)
+                    """
+                ),
+            )
+            self.assertEqual(
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py:898]: do we need this assertion?

I am ok with it... but it can cause future failure when we update part of the output.

File: src/towncrier/_git.py [hunk 4]
```diff
@@ -16,11 +28,23 @@ def remove_files(fragment_filenames, answer_yes):
     else:
         click.echo("I want to remove the following files:")
 
-    for filename in fragment_filenames:
+    for filename in sorted(fragment_filenames):
         click.echo(filename)
 
+    # Filter out files that are unknown to git
+    try:
+        known_fragments = run(
```
Comment (adiroiban) [inline src/towncrier/_git.py:36]: Maybe rename to something like this to hint that these are the files handled by git.

```suggestion
        git_fragments = run(
```

And shouldn't this be moved inside `if answer_yes` , to hint that this code is only needed if we need to delete the files.

And maybe we can resolve the path to absolute so that we don't have 2 variables.
git should be able to handle absolute paths.

File: src/towncrier/_git.py [hunk 5]
```diff
@@ -1,12 +1,24 @@
...
-
 import os
+import sys
+from subprocess import call, check_output, CalledProcessError
+
 import click
 
 
+def run(args, **kwargs):
+    # Use UTF-8 both when sys.stdout does not have .encoding (Python 2.7) and
```
Comment (adiroiban) [inline src/towncrier/_git.py:12]: Maybe have the comment as docstring ? 

And this is not really related to the git code.

Maybe keep is as the private `check._run` and update the private `_run` to always return decoded output.

Make sense?

File: src/towncrier/check.py [hunk 6]
```diff
@@ -8,15 +8,11 @@
...
 import click
 
-from subprocess import CalledProcessError, check_output, STDOUT
+from subprocess import CalledProcessError, STDOUT
 
 from ._settings import load_config_from_options
 from ._builder import find_fragments
-
-
-def _run(args, **kwargs):
```
Comment (adiroiban) [inline src/towncrier/check.py:17]: I would prefer to keep the private `_run` and refactor it to always return decoded output and add a docstring that explains what it does.

**Review by adiroiban — CHANGES_REQUESTED** [review]
Many thanks for the changes.

They look good.

I left a few minor comments regarding the documentation/docstrings of these changes.

I would like to keep the `_run` helper private and outside of the git code :)

Sorry for the late review.

I hope you will still have time to drive this PR.

Let me know if you are no longer interested in working on it and I will try to take over and finalize this PR.

Cheers

File: src/towncrier/test/test_build.py (hunk 3 above)
Comment (adiroiban) [reply src/towncrier/test/test_build.py:898]: And I think that this assertion is cause failures on Windows...as the path has a different separator.

**Comment by jaraco** [comment]
It looks like this issue has gotten a little stale, and since we haven't heard back from the contributor, I'm going to close it for now. If you'd like to revive the effort, feel free to ping this PR and we can re-open, or start a new one. Thanks.

**Comment by abompard** [comment]
Hey folks! Sorry I had totally forgotten about this PR. I'll try to find some time to finalize it, if it's still relevant.

File: src/towncrier/_git.py (hunk 5 above)
Comment (abompard) [reply src/towncrier/_git.py]: Yeah, but I need to call this function from the `remove_files` function, and it feels a bit weird to import a private method from the `check` module (which is a CLI module).
How would you do it?

**Comment by abompard** [comment]
Could you please reopen @jaraco ?

File: src/towncrier/_git.py [hunk 7]
```diff
@@ -1,12 +1,27 @@
...
+import sys
+from subprocess import call, check_output, CalledProcessError
+
 import click
 
 
+def run(args, **kwargs):
+    """Run a command.
+
+    Use UTF-8 both when sys.stdout does not have .encoding (Python 2.7) and
```
Comment (adiroiban) [inline src/towncrier/_git.py]: do we still support python 2.7 ?

I think that this comment, and maybe the implementation should be updated for py3 only.

**Comment by adiroiban** [comment]
Thanks for the update.

Can you please fix the conflicts and update the docstring for `run` to describe it's current behaviour

We no longer support python 2.7

**Comment by abompard** [comment]
I'm more than happy to remove Python 2.7 remnants :-D
Alright, it's updated and tests should pass now.

File: src/towncrier/_git.py [hunk 8]
```diff
@@ -5,12 +5,26 @@
...
-        call(["git", "rm", "--quiet"] + fragment_filenames)
+    if not fragment_filenames:
+        return
+
+    # Filter out files that are unknown to git
+    try:
+        git_fragments = check_output(
+            ["git", "ls-files"] + fragment_filenames, encoding="utf-8"
+        ).split("\n")
+        git_fragments = [os.path.abspath(f) for f in git_fragments if os.path.isfile(f)]
```
Comment (adiroiban) [inline src/towncrier/_git.py]: Minor comment

can we have this outside of try/except ?

Are we expecting this line to raise `CallerProcessError` ?

File: src/towncrier/_git.py [hunk 9]
```diff
@@ -5,12 +5,26 @@
...
+    except CalledProcessError:
```
Comment (adiroiban) [inline src/towncrier/_git.py:20]: can we have a comment describing in which case we exepct to have this error raised ?

File: src/towncrier/test/test_build.py [hunk 10]
```diff
@@ -747,8 +748,8 @@ def do_build_once_with(version, fragment_file, fragment):
                     "--yes",
                 ],
             )
-            # not git repository, manually remove fragment file
-            Path(f"newsfragments/{fragment_file}").unlink()
+            # fragment files unknown to git are removed even without a git repo
+            # Path(f"newsfragments/{fragment_file}").unlink()
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: I think that this needs a cleanup.

We should not merge commented code.

I think that we should repace this line with an assertion to check that the file is already removed.

File: src/towncrier/test/test_build.py [hunk 11]
```diff
@@ -845,8 +846,8 @@ def do_build_once_with(version, fragment_file, fragment):
                 ],
                 catch_exceptions=False,
             )
-            # not git repository, manually remove fragment file
-            Path(f"newsfragments/{fragment_file}").unlink()
+            # fragment files unknown to git are removed even without a git repo
+            # Path(f"newsfragments/{fragment_file}").unlink()
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: this needs cleanup

File: src/towncrier/test/test_build.py [hunk 12]
```diff
@@ -1530,3 +1531,55 @@ def test_orphans_in_non_showcontent_markdown(self, runner):
...
+        At build time, it will delete any fragment file regardless of its stage,
+        included files that are not part of the git reporsitory,
+        or are just staged or modified.
+        """
+        # 123 is committed, 124 is modified, 125 is just added, 126 is unknown
+
+        with open("foo/newsfragments/123.feature", "w") as f:
+            f.write("Adds levitation")
+        with open("foo/newsfragments/124.feature", "w") as f:
+            f.write("Extends levitation")
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py]: Maybe add more info to know the purpose of each file.

```suggestion
            f.write("Adds levitation. File committed.")
        with open("foo/newsfragments/124.feature", "w") as f:
            f.write("Extends levitation. File modified in git.")
```

File: src/towncrier/test/test_build.py [hunk 13]
```diff
@@ -1530,3 +1531,55 @@ def test_orphans_in_non_showcontent_markdown(self, runner):
...
+
+        commit()
+
+        with open("foo/newsfragments/125.feature", "w") as f:
+            f.write("Baz levitation")
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py]:
```suggestion
        with open("foo/newsfragments/125.feature", "w") as f:
            f.write("Baz levitation. Stagged file.")
```

File: src/towncrier/test/test_build.py [hunk 14]
```diff
@@ -1530,3 +1531,55 @@ def test_orphans_in_non_showcontent_markdown(self, runner):
...
+        with open("foo/newsfragments/126.feature", "w") as f:
+            f.write("Fix (literal) crash")
```
Comment (adiroiban) [inline src/towncrier/test/test_build.py]:
```suggestion
        with open("foo/newsfragments/126.feature", "w") as f:
            f.write("Fix (literal) crash. File outside of git.")
```

**Review by adiroiban — CHANGES_REQUESTED** [review]
Thanks for the changes.

I think that we are very close to being able to merge this.

But this needs more cleanup.

The commented lines in the test file are a blocker.

Thanks again.

**Comment by abompard** [comment]
I've made the changes, thanks a lot for your patience. FYI I do appreciate the in-depth reviews, even for minor style changes. I'm frequently on the other end of those ;-)
So thanks.

File: src/towncrier/newsfragments/357.feature.rst [hunk 15]
```diff
@@ -0,0 +1 @@
+It can now handle news fragments which are not part of the git repository. For example uncommited or unstagged files.
```
Comment (SmileyChris) [inline src/towncrier/newsfragments/357.feature.rst]:
```suggestion
``towncrier build`` now handles removing news fragments which are not part of the git repository. For example, uncommitted or unstaged files.
```

**Review by SmileyChris — CHANGES_REQUESTED** [review]
Currently, the command line docs say:
> By default, the processed news fragments are removed using git, which will also remove the fragments directory if now empty.

This should be updated. Maybe something like
> By default, the processed news fragments are removed. For any fragments committed in your git repository, ``git rm`` will be used (which will also remove the fragments directory if now empty).

**Comment by adiroiban** [comment]
Thanks for the update. I will let Chris do the final review and approval for this PR.

@abompard Note that we merge with squash.
There is not much use in rewriting the history for the PR branch.
Rewriting/amending the commits on the PR make it harder to do a followup for a review.

Cheers

**Comment by abompard** [comment]
> @abompard Note that we merge with squash. There is not much use in rewriting the history for the PR branch. Rewriting/amending the commits on the PR make it harder to do a followup for a review.

Ah, sorry about that, I didn't know.

FYI, I've had this same issue many times (trying to review PRs that rebase/rewrite) and I ended up writing a small tool to help me with the workflow: https://pypi.org/project/git-pr-branch/. The main idea is that it will download the different revisions of a PR in different branches that you can diff, even if it's always 1 commit that's being rebased.
Maybe you'll find it useful for your workflow as well.

**Comment by adiroiban** [comment]
No problem.
I am looking after `towncrier` as my 4th job ... so I don't have much time for it.
This is why, using the embedded GitHub web tools makes it easier to review.

For most of the reviews, I don't checkout the code on my computer.
I am just doing a code review and hope the automated tests are doing their job.

**Comment by adiroiban** [comment]
needs-review

**Review by adiroiban — APPROVED** [review]
This looks good to me.
I will wait a bit more for Chris to review this.

If there is a big delay, we can have this merged.

There is no rush, as there is no relesed planed anytime soon...
but we might want do to a new release soon :)

**Comment by adiroiban** [comment]
I think that the comments from Chris were address.
I have merged this.

Thanks for your help!
